packages/shopping-cart/src/shopping-cart-manager.ts

```diff
@@ -0,0 +1,170 @@
+import debugFactory from 'debug';
+import {
+	getShoppingCartManagerState,
+	createSubscriptionManager,
+	createLastValidResponseCartManager,
+	createActionPromisesManager,
+	noopManager,
+} from './managers';
+import { createActions } from './shopping-cart-actions';
+import {
+	getInitialShoppingCartState,
+	reducerWithQueue as shoppingCartReducer,
+} from './shopping-cart-reducer';
+import { createTakeActionsBasedOnState } from './state-based-actions';
+import { createCartSyncManager } from './sync';
+import type {
+	GetCart,
+	SetCart,
+	ShoppingCartManagerClient,
+	ShoppingCartManagerState,
+	ShoppingCartManager,
+	ShoppingCartAction,
+	ResponseCart,
+	ShoppingCartReducerDispatch,
+	DispatchAndWaitForValid,
+	ActionPromises,
+	ShoppingCartState,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:shopping-cart-manager' );
+
+function createDispatchAndWaitForValid(
+	dispatch: ShoppingCartReducerDispatch,
+	actionPromises: ActionPromises
+): DispatchAndWaitForValid {
+	return ( action ) => {
+		return new Promise< ResponseCart >( ( resolve ) => {
+			actionPromises.add( resolve );
+			dispatch( action );
+		} );
+	};
+}
+
+function createShoppingCartManager(
+	cartKey: string,
+	getCart: GetCart,
+	setCart: SetCart
+): ShoppingCartManager {
+	let state = getInitialShoppingCartState();
+
+	const subscriptionManager = createSubscriptionManager( cartKey );
+
+	// When an action is dispatched that modifies the cart (eg:
+	// addProductsToCart), the reducer modifies the `responseCart` stored in
+	// state. That (incomplete) data is then sent off to the server to be
+	// validated and filled-in with additional properties before being returned
+	// to the ShoppingCartManager. Because of this, optimistic updating of the
+	// cart is not possible (it may change significantly in that round trip) so
+	// we don't want to return the raw `responseCart` to the consumer. Instead,
+	// we keep a copy of the `responseCart` the last time the state had a `valid`
+	// CacheStatus and pass that to our consumers. The consumers can use
+	// `isPendingUpdate` to know when the cart data is updating.
+	const lastValidResponseCart = createLastValidResponseCartManager( state );
+
+	const syncManager = createCartSyncManager( cartKey, getCart, setCart );
+	const actionPromises = createActionPromisesManager();
+	const takeActionsBasedOnState = createTakeActionsBasedOnState( syncManager );
+
+	let deferredStateCheck: ReturnType< typeof setTimeout >;
+	const dispatch = ( action: ShoppingCartAction ) => {
+		debug( `dispatching action for cartKey ${ cartKey }`, action.type );
+		const newState = shoppingCartReducer( state, action );
+		if ( newState === state ) {
+			if ( ! isStatePendingUpdateOrQueuedAction( state ) ) {
+				debug( 'state did not change; resolving action promises' );
+				actionPromises.resolve( state.responseCart );
+			}
```

Comment:
Isn't this check redundant? If the `state` hasn't changed, then the promises resolution must have been done by the previous update that changed it. At this moment, `actionPromises` are guaranteed to be empty, aren't they?

I can see that this would be needed in case of:
```js
await dispatchAndWaitForValid( actionThatDoesntChangeState )
```
but are there such actions that don't change state?

Reply:
> but are there such actions that don't change state?

There are some, like changing the location to an identical value, replacing a product with the same product, adding a coupon that's already applied, or removing a coupon when no coupon is applied. I added a test for one of them and this block is required for it to pass. They probably don't come up that often but having Promises that don't resolve felt wrong to me.

https://github.com/Automattic/wp-calypso/blob/d80882288abe623a8ff54c3084b3060242aff238/packages/shopping-cart/test/cart-manager.ts#L108

Reply:
I see, so we want to resolve action promises on every state update, even a noop one, and to take actions (i.e., sync) and notify listeners only on a real change, right?

The `dispatch` logic can be reorganized a bit to call `actionPromises.resolve()` only once at the right place. See #55703 where I tried to do that.